plugins/fileservices/fileservices.cpp

```diff
@@ -198,63 +198,123 @@ static IPropertyTree *getEnvironmentTree(IConstEnvironment * daliEnv)
     return getHPCCEnvironment();
 }
 
-static StringAttr espurl;           // Default ESP url if none specified
+static void setServerAccess(CClientFileSpray &server, IConstWorkUnit * wu)
+{
+    StringBuffer user, password, token;
+    wu->getSecurityToken(StringBufferAdaptor(token));
+    extractToken(token.str(), wu->queryWuid(), StringBufferAdaptor(user), StringBufferAdaptor(password));
+    server.setUsernameToken(user.str(), password.str(), "");
+}
+
+static StringArray availableWsFS;
 static CriticalSection espURLcrit;
 
-static const char *getEspServerURL(const char *param)
+static void addConfiguredWsFSUrl(const char * url)
+{
+    CriticalBlock b(espURLcrit);
+    availableWsFS.appendUniq(url);
+}
+
+static bool contactWsFS(const char * espurl, IConstWorkUnit * wu)
+{
+    CClientFileSpray server;
+    server.addServiceUrl(espurl);
+    setServerAccess(server, wu);
+
+    try
+    {
+        Owned<IClientEchoDateTime> req = server.createEchoDateTimeRequest();
+        Owned<IClientEchoDateTimeResponse> result = server.EchoDateTime(req);
+
+        if (result->getResult())
+            return true;
+    }
+    catch(...)
+    {
+        PROGLOG("Could not contact WsFS: '%s'",espurl);
+    }
+    return 0;
+}
+
+static const char * getNextAliveWsFSURL(IConstWorkUnit * wu)
+{
+    CriticalBlock b(espURLcrit);
+
+    for (int index = 0; index < availableWsFS.length(); index++)
+    {
+        const char * currentUrl = availableWsFS.item(index);
+        if (contactWsFS(currentUrl, wu))
+            return currentUrl;
+        //else remove it from the list?
```

Comment:
Should NOT remove from the list - if a user has no rights on a given ESP it does not mean that other users don't (and this list is static so shared by all users)